invokeai/app/services/batch_manager.py

```diff
@@ -0,0 +1,139 @@
+import networkx as nx
+import copy
+
+from abc import ABC, abstractmethod
+from itertools import product
+from pydantic import BaseModel, Field
+from fastapi_events.handlers.local import local_handler
+from fastapi_events.typing import Event
+
+from invokeai.app.services.events import EventServiceBase
+from invokeai.app.services.graph import Graph, GraphExecutionState
+from invokeai.app.services.invoker import Invoker
+from invokeai.app.services.batch_manager_storage import (
+    BatchProcessStorageBase,
+    BatchSessionNotFoundException,
+    Batch,
+    BatchProcess,
+    BatchSession,
+    BatchSessionChanges,
+)
+
+
+class BatchManagerBase(ABC):
+    @abstractmethod
+    def start(self, invoker: Invoker):
+        pass
+
+    @abstractmethod
+    def create_batch_process(self, batches: list[Batch], graph: Graph) -> str:
+        pass
+
+    @abstractmethod
+    def run_batch_process(self, batch_id: str):
+        pass
+
+    @abstractmethod
+    def cancel_batch_process(self, batch_process_id: str):
+        pass
+
+
+class BatchManager(BatchManagerBase):
+    """Responsible for managing currently running and scheduled batch jobs"""
+
+    __invoker: Invoker
+    __batches: list[BatchProcess]
+    __batch_process_storage: BatchProcessStorageBase
+
+    def __init__(self, batch_process_storage: BatchProcessStorageBase) -> None:
+        super().__init__()
+        self.__batch_process_storage = batch_process_storage
+
+    def start(self, invoker: Invoker) -> None:
+        # if we do want multithreading at some point, we could make this configurable
+        self.__invoker = invoker
+        self.__batches = list()
+        local_handler.register(event_name=EventServiceBase.session_event, _func=self.on_event)
+
+    async def on_event(self, event: Event):
+        event_name = event[1]["event"]
+
+        match event_name:
+            case "graph_execution_state_complete":
+                await self.process(event, False)
+            case "invocation_error":
+                await self.process(event, True)
+
+        return event
+
+    async def process(self, event: Event, err: bool):
+        data = event[1]["data"]
+        batch_session = self.__batch_process_storage.get_session(data["graph_execution_state_id"])
+        if not batch_session:
+            return
+        updateSession = BatchSessionChanges(
+            state='error' if err else 'completed'
+        )
+        batch_session = self.__batch_process_storage.update_session_state(
+            batch_session.batch_id,
+            batch_session.session_id,
+            updateSession,
+        )
+        self.run_batch_process(batch_session.batch_id)
+
+    def _create_batch_session(self, batch_process: BatchProcess, batch_indices: list[int]) -> GraphExecutionState:
+        graph = copy.deepcopy(batch_process.graph)
+        batches = batch_process.batches
+        g = graph.nx_graph_flat()
+        sorted_nodes = nx.topological_sort(g)
+        for npath in sorted_nodes:
```

Comment:
So is `batch_id` just added in the item json somewhere in DB, versus an additional column?

Reply:
`batch_id` is a primary key on `batch_process`, and a foreign key on `batch_session`